stl/inc/format

```diff
@@ -0,0 +1,363 @@
+        throw _STD format_error("Invalid integer");
+    }
+    if(_Integer < 0) {
+        throw _STD format_error("Nonnegative integer is nonnonnegative");
+    }
+    return _Ptr;
+}
+
+
+
+template<typename _CharT, typename _Callbacks_type> requires _Parse_arg_id_callbacks<_Callbacks_type, _CharT>
+constexpr const _CharT* _Parse_arg_id(const _CharT* _Begin, const _CharT* _End, _Callbacks_type&& _Callbacks) {
+    assert(_Begin != _End);
+    _CharT _Ch = *_Begin;
+    // No id provided, format string is using automatic indexing.
+    if(_Ch == '}' || _Ch == ':') {
+        _Callbacks._On_auto_id();
+        return _Begin;
+    }
+    if(_Ch >= '0' && _Ch <= '9') {
+        int _Index = 0;
+        // arg_id is not allowed to have any leading zeros, but is allowed to be
+        // equal to zero (but not '00'). So if _Ch is zero we skip the parsing, leave
+        // _Index set to zero and let the validity checks below ensure that the arg_id
+        // wasn't something like "00", or "023".
+        if( _Ch != '0') {
+            _Begin = _Parse_nonnegative_integer(_Begin, _End, _Index);
+        } else {
+            ++_Begin;
+        }
+        // The format string shouldn't end right after the index number.
+        // The only things permitted after the index are the end of the replacement field ('}')
+        // or the beginning of the format spec (':').
+        if (_Begin == _End || (*_Begin != '}' && *_Begin != ':')) {
+            throw _STD format_error("invalid format string");
+        }
+        _Callbacks._On_manual_id(_Index);
+        return _Begin;
+    }
+    // This is where we would parse named arg ids if std::format were to support them.
+    throw _STD format_error("Invalid format string.");
+}
+
+template <class _CharT, typename _Callbacks_type> requires _Parse_spec_callbacks<_Callbacks_type, _CharT>
+constexpr const _CharT* _Parse_align(const _CharT* _Begin, const _CharT* _End, _Callbacks_type&& _Callbacks) {
+    // done with parsing, or reached the end of a replacement field.
+    if (_Begin == _End || *_Begin == '}') {
+        return _Begin;
+    }
+
+    // align and fill
+    auto _Align    = _Align::_None;
+    auto _Align_pt = _Begin + 1;
+    if (_Align_pt == _End) {
+        _Align_pt = _Begin;
+    }
+    if (_Begin == _End) {
+        _Begin--;
+    }
+    for (;;) {
+        switch (static_cast<char>(*_Align_pt)) {
+        case '<':
+            _Align = _Align::_Left;
+            break;
+        case '>':
+            _Align = _Align::_Right;
+            break;
+        case '^':
+            _Align = _Align::_Center;
+            break;
+        }
+        if (_Align != _Align::_None) {
+            if (_Align_pt != _Begin) {
+                if (*_Begin == '{') {
+                    throw format_error("invalid fill character '{'");
+                }
+                _Callbacks._On_fill({_Begin, static_cast<size_t>(_Align_pt - _Begin)});
+                _Begin = _Align_pt + 1;

```

Comment:
....in which case this is equivalent to `++_Begin`, so we could drop this line and move 142 out of the conditional.

Reply:
The reason it's set up like this is because the fill character may want to be multi-byte (which I need to handle, and figure out what that means for wide characters), in which case line 114 may set _Align_pt to more than 1 past _Begin